Generalise magic sets for aggregation and negation

demianw opened the pull request.

jonasrenault commented:
I think this is good for a review @demianw

demianw commented on neurolang/datalog/magic_sets.py:
```diff
@@ -68,12 +199,122 @@ def magic_rewrite(query, datalog):
+        the set of magic initialization rules
+    """
+    magic_init_rules = []
+    for predicate in extract_logic_predicates(adorned_query.antecedent):
```
This will iterate you the predicates that are of the form `P(x, y, z, ...)` or not `P(x, y, z, ...)`. What happens if you find a negated predicate? Is it verified before that rules have no negation? Also you might get builtins, i.e., `a == b`. Should you also avoid adding facts for these in the init rules?

Or this is only called with the adorned query which has avoided all these?

demianw commented on neurolang/datalog/magic_sets.py:
```diff
@@ -68,12 +199,122 @@ def magic_rewrite(query, datalog):
+        consequent = rule.consequent
+        magic_head = magic_predicate(consequent)
+        body_predicates = (magic_head,)
+        for predicate in extract_logic_predicates(rule.antecedent):
```
Same questions as in line 257. Not all predicates have functors (case of negation)

demianw commented on neurolang/datalog/tests/test_magic_sets.py:
```diff
@@ -15,7 +23,10 @@
+    TranslateToLogicWithAggregation,
+    BuiltinAggregationMixin,
+    DatalogWithAggregationMixin,
+    DatalogProgramNegationMixin,
```
Maybe add a couple tests to see what happens when Negation and builtins (constant functions as predicates, like `operator.eq`)  are present?

demianw reviewed (commented):
Good job @jonasrenault . Some minor changes are needed I think.

jonasrenault commented:
I've update the code to raise an exception if there exists a negated predicate in the rules, since the algorithm does not (yet) support negations.

I've also changed the way the `query_init_rules` are created to initialise all predicates that have a constant argument and not just those that are in the query rule. This allows to propagate constants that are defined inside a program rule and not just those in the query rule.

I've added tests for these two cases. Tests for builtins already existed.

demianw commented:
> I've update the code to raise an exception if there exists a negated predicate in the rules, since the algorithm does not (yet) support negations.
> 
> I've also changed the way the `query_init_rules` are created to initialise all predicates that have a constant argument and not just those that are in the query rule. This allows to propagate constants that are defined inside a program rule and not just those in the query rule.
> 
> I've added tests for these two cases. Tests for builtins already existed.

Great!
